### DISCUSSION TITLE
Does terratest support snapshot based assertion?

### DISCUSSION THREAD

**Discussion by jguionnet** [gruntwork-io/knowledge-base#780] [Help]
[helm-unittest](https://github.com/helm-unittest/helm-unittest) as a great feature called [snapshot testing](https://github.com/helm-unittest/helm-unittest#snapshot-testing ). 

Sometimes you may just want to check that the rendered manifest did not change between helm package version without every details asserted.

With Terratest, we could add a feature in helm module to support that. For example, we could

- Add a set of helper functions, like the `UnmarshalK8SYaml` functions. 
  - one function would create/refresh the local snapshot
  - one function would diff the manifest generated by the test against the local snapshot. It would return false if they are different and would have an option to display the difference in that case. 

---

<ins datetime="2023-11-04T00:01:02Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110569">Tracked in ticket #110569</a></p>
</ins>

**Comment by gitsstewart** [comment]
@jguionnet We will take a look at your suggestion and see if it fits on our current priority list for terratest. In the meantime, we welcome you to submit a pull request for the community to review.

**Comment by denis256** [comment]
Hello,
yes, snapshot testing functions will be helpful, looking forward for PRs :)

**Reply by jguionnet** [reply]
Created an issue for it https://github.com/gruntwork-io/terratest/issues/1377

**Reply by so0k** [reply]
looks like this was closed for k8s yaml with https://github.com/gruntwork-io/terratest/pull/1378

for reference of future visitors/googlers, golang has many snapshot testing libraries for example:
- github.com/google/go-cmp/cmp
- github.com/gkampitakis/go-snaps
 
 both work great, in my case I wanted to be able to template and test a fragment, so I built some helpers
 
 https://github.com/envtio/base/blob/v0.0.7/integ/aws/iam/helpers_test.go#L32
 
 However, as I try to match the Integration testing behaviour of AWS CDK, I created JMESPath based regex assertions instead

for example:
 - https://github.com/envtio/base/blob/v0.0.7/integ/aws/edge/edge_test.go#L70-L93
 - https://github.com/envtio/base/blob/v0.0.7/integ/aws/compute/function_test.go#L89-L106

Not sure what's the best place to discuss if there's interest in upstreaming some of this functionality, but I felt it was related to snapshot testing and assertions, and this was the only discussion thread with recent activity on terratest.

EDIT 3: As I was unaware there was pre-existing work for snapshot testing in terratest (maybe it should be outside of helm and more under a module of its own?

I'll have a look at the existing libraries used instead

- github.com/gonvenience/ytbx
- github.com/homeport/dyff/pkg/dyff
